### PR TITLE
Bug dot paths

### PR DESCRIPTION
**Pull request by parkerpersaud**
https://trello.com/c/2zKgPQhh/97-375-fix-requests-with-dots-404

### CODE REVIEW

File: src/main/java/me/karakelley/http/exceptions/AttemptedDirectoryTraversalException.java [hunk 1]
```diff
@@ -0,0 +1,7 @@
+package me.karakelley.http.exceptions;
+
+public class AttemptedDirectoryTraversalException extends Exception {
```
Comment (mikeknep) [inline src/main/java/me/karakelley/http/exceptions/AttemptedDirectoryTraversalException.java]: Any particular reason this one is checked, and not a `RuntimeException` like your others? (You recently made all your others unchecked, so just wondering why this one is different.)

File: src/main/java/me/karakelley/http/server/HttpRequestReader.java [hunk 2]
```diff
@@ -78,8 +81,13 @@ private String readRequestLineAndHeaders() throws IOException {
     return buffer.toByteArray();
   }
 
-  private String validPath(String path) {
-    if (path == null) throw new InvalidRequestException("Invalid Path!");
+  private String validPath(String path) throws AttemptedDirectoryTraversalException {
+    Path userPath = Paths.get(path);
+
+    Path normalizedPath = userPath.normalize();
+    if (normalizedPath != userPath || userPath.startsWith("..")) {
```
Comment (mikeknep) [inline src/main/java/me/karakelley/http/server/HttpRequestReader.java]: 1. Can you walk me through how `normalize` and this comparison works? I'm guessing `normalize` does something like:
```
Paths.get("/Users/mikeknepper/projects/server/../../projects").normalize()
# => "/Users/mikeknepper/projects" (as a Path)
```
Does the comparison then check that they end up in the same place?

2. Could someone get around this check by requesting `/someDirectory/../../.ssh` ? (Do you possibly need to check if `normalizedPath` starts with dots instead of `userPath`?)

**Review by mikeknep — COMMENTED** [review]
A couple questions, but overall this PR is really nice—it's tightly focused on a specific bug making it easily digestible, and it provides tests to ensure the bug does not reoccur after potential future refactoring. I would prefer seeing the tests and the production code together in a single commit (I might argue this whole PR could be a single commit) but in general very nice job!

File: src/main/java/me/karakelley/http/exceptions/AttemptedDirectoryTraversalException.java (hunk 1 above)
Comment (parkerpersaud) [reply src/main/java/me/karakelley/http/exceptions/AttemptedDirectoryTraversalException.java]: I wanted to be more explicit with this particular exception. I'm not sure if that's a valid enough answer but I'm still going back and forth on pros/cons on checked vs unchecked.

File: src/main/java/me/karakelley/http/server/HttpRequestReader.java (hunk 2 above)
Comment (parkerpersaud) [reply src/main/java/me/karakelley/http/server/HttpRequestReader.java]: Yes, that's how `normalize` works - So once the path has been normalized and it's different that what you requested then it'll throw an `AttemptedDirectoryTraversalException`. 

So if you requested `/somedirectory/../../.ssh` it'll be normalized to /somedirectory/.ssh. Then we check to see if `/somedirectory/../../.ssh` is equal to ` /somedirectory/.ssh`. They're not equal so exception is thrown. 

I believe checking either `userPath` or `normalizedPath` for preceding dots will the same since `normalize()` doesn't remove any preceding dots however it does seem more readable to check the `normalizedPath` instead.

File: src/main/java/me/karakelley/http/filesystem/PublicDirectory.java [hunk 3]
```diff
@@ -35,17 +35,13 @@ private PublicDirectory(String path, FileFinder fileFinderCache) {
...
   public List<File> getDirectoriesAndFiles(String requestedResource) {
     File directory = getPath(requestedResource).toFile();
-    if (resourceExists(requestedResource)) {
-      return Arrays.stream(directory.listFiles())
-              .filter(file -> file.isDirectory() || file.isFile())
-              .collect(Collectors.toList());
-    } else {
-      return new ArrayList<>();
-    }
+    return Arrays.stream(directory.listFiles())
```
Comment (mikeknep) [inline src/main/java/me/karakelley/http/filesystem/PublicDirectory.java]: Always give props for removing an `if/else`!

File: src/main/java/me/karakelley/http/filesystem/PublicDirectory.java [hunk 4]
```diff
@@ -91,10 +87,22 @@ public Path getPath(String requestedResource) {
 
   public void updateFileContents(String path, byte[] contents) {
     File newFile = getPath(path).toFile();
-    try (FileOutputStream outputStream = new FileOutputStream(newFile, false)) {
+    try(FileOutputStream outputStream = new FileOutputStream(newFile, false)) {
```
Comment (mikeknep) [inline src/main/java/me/karakelley/http/filesystem/PublicDirectory.java]: Ultra nitpick: your diff adds a space after one `try` (76) and removes a space for another `try` (here).

File: src/main/java/me/karakelley/http/handlers/StaticFilesHandler.java [hunk 5]
```diff
@@ -28,7 +28,9 @@ public Response respond(Request request) {
     if (publicDirectory.resourceExists(requestedResource)) {
       response = new Ok();
       return serveResponse(response, requestedResource);
-    } else return new NotFound();
+    } else {
+      return new NotFound();
```
Comment (mikeknep) [inline src/main/java/me/karakelley/http/handlers/StaticFilesHandler.java:32]: Thank you! Again it's a nit, but this style is much more idiomatic.

File: src/test/java/me/karakelley/http/publicdirectory/PublicDirectoryTest.java [hunk 6]
```diff
@@ -76,13 +75,6 @@ void testResourceDoesntExist() {
     assertFalse(publicDirectory.resourceExists("/someBadPath"));
   }
 
-  @Test
-  void testNoFilesInDirectory() {
```
Comment (mikeknep) [inline src/test/java/me/karakelley/http/publicdirectory/PublicDirectoryTest.java:80]: Out of curiosity, why did you remove this test?

File: src/main/java/me/karakelley/http/filesystem/PublicDirectory.java [hunk 7]
```diff
@@ -91,10 +87,22 @@ public Path getPath(String requestedResource) {
...
       outputStream.write(contents);
     } catch (IOException e) {
       throw new RuntimeException(e);
     }
   }
+
+  public boolean isValidPath(String requestedPath) {
```
Comment (mikeknep) [inline src/main/java/me/karakelley/http/filesystem/PublicDirectory.java]: Does any external client call this method? Does it need to be public? Or is it only public for purposes of testing?

File: src/test/java/me/karakelley/http/publicdirectory/PublicDirectoryTest.java [hunk 8]
```diff
@@ -197,4 +187,19 @@ void testUpdateContentsOfFile() {
...
+  @Test
+  void testValidPath() {
+    TempFilesHelper.withTempDirectory(directory ->  {
+      PublicDirectory publicDirectory = PublicDirectory.create(directory.toString(), new FileFinderCache(new RealFileFinder()));
+      assertEquals(true, publicDirectory.isValidPath("testing/../testing"));
+    });
+  }
+
+  @Test
+  void testInvalidPath() {
```
Comment (mikeknep) [inline src/test/java/me/karakelley/http/publicdirectory/PublicDirectoryTest.java:199]: Follow-up on the public-vs-private `isValidPath` comment:
On the one hand, I like these two tests here because they help document the behavior of `isValidPath` method, which I personally find a little hard to follow (but probably mainly because it uses a lot of std lib methods with which I'm not super familiar). On the other hand, the 404-on-attempted-dir-traversal test does cover the public-facing behavior at a nice, high level. Maybe there's a compromise, keeping the 404 test and replacing these with something just a little higher up in `PublicDirectory`?

File: src/test/java/me/karakelley/http/publicdirectory/PublicDirectoryTest.java (hunk 6 above)
Comment (parkerpersaud) [reply src/test/java/me/karakelley/http/publicdirectory/PublicDirectoryTest.java:80]: I removed this particular test because logically you'd never get to this point in the program if the resource didn't exist (the naming was poor and doesn't express this) now that the validation is all done at a higher level. A test similar to this does exist at a higher level as well.

File: src/test/java/me/karakelley/http/publicdirectory/PublicDirectoryTest.java (hunk 8 above)
Comment (parkerpersaud) [reply src/test/java/me/karakelley/http/publicdirectory/PublicDirectoryTest.java:199]: The method should be private. I initially had a different implementation and when I changed it to be called within `resourceExists()` instead I forgot to change it to private. I could easily test this through `resourceExists()` which is probably more appropriate entry point for testing it.

**Review by mikeknep — APPROVED** [review]

File: src/main/java/me/karakelley/http/filesystem/PublicDirectory.java [hunk 9]
```diff
@@ -97,4 +93,24 @@ public void updateFileContents(String path, byte[] contents) {
       throw new RuntimeException(e);
     }
   }
+
+  private boolean isValidPath(String requestedPath) {
+    File canonicalPath = normalizeFullPath(requestedPath);
+
+    return canonicalPath.toString().contains(normalizedDocumentRoot().toString());
```
Comment (aaronlahey) [inline src/main/java/me/karakelley/http/filesystem/PublicDirectory.java]: It feels strange to be checking if a path is valid by doing a string comparison. Can you use [this](https://docs.oracle.com/javase/7/docs/api/java/nio/file/Path.html#startsWith(java.nio.file.Path))?

File: src/main/java/me/karakelley/http/handlers/StaticFilesHandler.java (hunk 5 above)
Comment (aaronlahey) [reply src/main/java/me/karakelley/http/handlers/StaticFilesHandler.java:32]: 💯

**Review by aaronlahey — APPROVED** [review]
Approved with my one nitpick addressed.

File: src/main/java/me/karakelley/http/filesystem/PublicDirectory.java (hunk 9 above)
Comment (parkerpersaud) [reply src/main/java/me/karakelley/http/filesystem/PublicDirectory.java]: So if I pass in the directory `./thispath/somewhere` the canonical path is something like `/Users/karakelley/code/thispath/somewhere` and `normalizedDocumentRoot` would be `/thispath/somewhere` which means the canonical path would never start with `normalizedDocumentRoot` (because its missing part of the full path). I really just want to see if the full document root is included within the requested path. I can see if there might be another alternative to string comparison but so far not seeing anything in the `Path` interface.
